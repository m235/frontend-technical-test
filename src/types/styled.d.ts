import 'styled-components'

declare module 'styled-components' {
  export interface DefaultTheme {
    palette: {
      primary: string
      secondary: string
      backgrounds: {
        light: string
        contrasted: string
      }
      grays: {
        lighter: string
        ligth: string
        normal: string
        dark: string
        darker: string
      }
      texts: {
        dark: string
        light: string
      }
    }
    spacing: {
      s050: string
      s100: string
      s200: string
      s250: string
      s300: string
      s400: string
    }
    elevation: {
      z0: string
      z1: string
    }
    breakpoint: {
      tablet: string
      desktop: string
    }
  }
}
