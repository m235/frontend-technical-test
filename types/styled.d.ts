import 'styled-components'

declare module 'styled-components' {
  export interface DefaultTheme {
    palette: {
      primary: string
      secondary: string
      backgrounds: {
        light: string
        contrasted: string
      }
      texts: {
        dark: string
        light: string
      }
    }
    spacing: {
      s050: string
      s100: string
      s200: string
      s250: string
      s300: string
      s400: string
    }
    elevation: {
      z0: string
      z1: string
    }
    breakpoint: {
      tablet: string
      desktop: string
    }
  }
}
