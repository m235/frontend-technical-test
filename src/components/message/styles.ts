import styled, { css } from 'styled-components'

const mineStyle = css`
  justify-content: flex-end;
  align-items: flex-end;
  border-radius: 8px;
  
  background-color: ${({ theme }) => theme.palette.primary};
  color: #ffffff;
`

const theirsStyle = css`
  justify-content: flex-start;
  align-items: flex-start;
  
  border-radius: 8px;
  
  background-color: ${({ theme }) => theme.palette.backgrounds.light};
  color: rgb(26, 26, 26);
`

export const MessageContainer = styled.div<{ isMine: boolean }>`
  display: flex;
  padding: ${({ theme }) => theme.spacing.s200};
  
  font-size: 14px;
  ${({ isMine }) => (isMine ? mineStyle : theirsStyle)};
  word-break: break-word;
`
