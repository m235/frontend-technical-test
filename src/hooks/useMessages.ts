import type { AxiosError } from 'axios'
import { useQuery } from 'react-query'

import { getMessages, MessagesApi } from '@/services/messages'
import type { Message } from '@/types/message'

const useMessages = (conversationId: number) => {
  return useQuery<Message[], AxiosError>([MessagesApi.MESSAGES, conversationId], () => getMessages(conversationId), {
    enabled: !!conversationId,
    useErrorBoundary: true,
  })
}

export default useMessages
