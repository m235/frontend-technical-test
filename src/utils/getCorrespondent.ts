import type { Conversation } from '@/types/conversation'

export const getCorrespondent = (userId: number, conversation: Conversation) => {
  switch (userId) {
    case conversation.senderId:
      return conversation.recipientNickname
    case conversation.recipientId:
      return conversation.senderNickname
    default:
      throw new Error('You are nor the recipient nor the sender')
  }
}
