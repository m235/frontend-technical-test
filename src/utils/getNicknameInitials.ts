export const getNicknameInitials = (nickname: string) =>
  nickname
    .split(' ')
    .map((string) => string.trim().charAt(0))
    .slice(0, 2)
    .join('')
