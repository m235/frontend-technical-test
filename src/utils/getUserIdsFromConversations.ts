import type { Conversation } from '@/types/conversation'

export const getUserIdsFromConversations = (conversations?: Conversation[]) =>
  Array.from<number>(new Set(conversations?.reduce((acc, curr) => [...acc, curr.recipientId, curr.senderId], []) ?? []))
