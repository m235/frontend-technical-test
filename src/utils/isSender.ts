import type { Conversation } from '@/types/conversation'

export const isSender = (userId: number, conversation: Conversation) => userId === conversation.senderId
