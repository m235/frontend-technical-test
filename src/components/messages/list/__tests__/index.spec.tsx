import MessagesList from '@/components/messages/list'
import useAuth from '@/hooks/useAuth'
import useMessages from '@/hooks/useMessages'
import { render } from '@/tests/utils'
import type { Conversation } from '@/types/conversation'
import { Message } from '@/types/message'

const mockedUseAuth = useAuth as jest.Mock<any>
const mockedUseMessages = useMessages as jest.Mock<any>
jest.mock('@/hooks/useAuth')
jest.mock('@/hooks/useMessages')

describe('components:conversations:list', () => {
  const conversation: Conversation = {
    id: 1,
    recipientNickname: 'Johnny Hallyday',
    recipientId: 2,
    senderNickname: 'John Travolta',
    senderId: 1,
  }
  const messages: Message[] = [
    { id: 1, timestamp: new Date().getTime(), authorId: 1, conversationId: 1, body: 'Foo' },
    { id: 2, timestamp: new Date().getTime(), authorId: 2, conversationId: 1, body: 'Bar' },
  ]

  beforeEach(() => {
    mockedUseAuth.mockImplementation(() => ({ userId: 1, user: [{ id: 1, nickname: 'John Travolta', token: 'x' }] }))
    mockedUseMessages.mockImplementation(() => ({ isLoading: true }))
  })
  afterEach(() => {
    jest.clearAllMocks()
  })
  it('should render default correctly', () => {
    const { container } = render(<MessagesList conversationId={conversation.id} />)

    expect(container).toMatchSnapshot()
  })
  it('should render loading messages', () => {
    const { getByTestId } = render(<MessagesList conversationId={conversation.id} />)

    expect(getByTestId('message-loading')).toBeInTheDocument()
  })
  it('should render no message', () => {
    mockedUseMessages.mockImplementation(() => ({ isLoading: false, data: [] }))
    const { getByTestId } = render(<MessagesList conversationId={conversation.id} />)

    expect(getByTestId('no-message')).toBeInTheDocument()
  })
  it('should render all the messages', () => {
    mockedUseMessages.mockImplementation(() => ({ isLoading: false, data: messages }))
    const { getByText } = render(<MessagesList conversationId={conversation.id} />)

    expect(getByText('Foo')).toBeInTheDocument()
    expect(getByText('Bar')).toBeInTheDocument()
  })
})
