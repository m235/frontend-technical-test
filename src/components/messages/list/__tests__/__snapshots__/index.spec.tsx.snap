// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components:conversations:list should render default correctly 1`] = `
<div>
  <div
    class="sc-hKMtZM gbmWKO"
  >
    <div
      class="sc-jSMfEi cKQEPY"
      data-testid="message-loading"
    >
      <i
        class="fa-solid fa-message"
      />
      <p
        class="sc-gKXOVf gLOqyj"
      >
        Loading messages...
      </p>
    </div>
  </div>
</div>
`;
